nvflare/app_common/streamers/file_downloader.py

```diff
@@ -0,0 +1,230 @@
+            file_name: name of the file to be downloaded
+            file_downloaded_cb: CB to be called when the file is done downloading
+            **cb_kwargs: args to be passed to the CB
+
+        Returns: reference id for the file.
+
+        The file_downloaded_cb must follow this signature:
+
+            cb(ref_id: str, to_site: str, status: str, file_name: str, **cb_kwargs)
+
+        """
+        obj = _File(file_name)
+        return ObjDownloader.add_download_object(
+            transaction_id=transaction_id,
+            obj=obj,
+            obj_downloaded_cb=cls._file_downloaded,
+            app_downloaded_cb=file_downloaded_cb,
+            **cb_kwargs,
+        )
+
+    @classmethod
+    def _file_downloaded(cls, ref_id: str, to_site: str, status: str, obj: _File, app_downloaded_cb, **cb_kwargs):
+        if app_downloaded_cb:
+            app_downloaded_cb(ref_id, to_site, status, obj.name, **cb_kwargs)
+
+
+class _ChunkConsumer(Consumer):
+
+    def __init__(self, location: str):
+        Consumer.__init__(self)
+        self.location = location
+        self.file_path = os.path.join(location, str(uuid.uuid4()))
+        self.file = open(self.file_path, "wb")
+        self.total_bytes = 0
+        self.error = None
+
+    def consume(self, ref_id, state: dict, data: Any) -> dict:
+        assert isinstance(data, bytes)
+        self.file.write(data)
+        self.total_bytes += len(data)
+        return {_StateKey.RECEIVED_BYTES: self.total_bytes}
+
+    def download_failed(self, ref_id, reason: str):
+        self.logger.error(f"failed to download file with ref {ref_id}: {reason}")
+        self.error = reason
+        self.file.close()
+
+    def download_completed(self, ref_id: str):
+        self.file.close()
+
+
+def download_file(
+    from_fqcn: str,
+    ref_id: str,
+    per_request_timeout: float,
+    cell: Cell,
+    location: str = None,
+    secure=False,
+    optional=False,
+    abort_signal=None,
+) -> (str, Optional[str]):
+    """Download the referenced file from the file owner.
+
+    Args:
+        from_fqcn: FQCN of the file owner.
+        ref_id: reference ID of the file to be downloaded.
+        per_request_timeout: timeout for requests sent to the file owner.
+        cell: cell to be used for communicating to the file owner.
+        location: dir for keeping the received file. If not specified, will use temp dir.
+        secure: P2P private mode for communication
+        optional: supress log messages of communication
+        abort_signal: signal for aborting download.
+
+    Returns: tuple of (error message if any, full path of the downloaded file).
+
+    """
+    if location is not None:
+        if not os.path.exists(location):
+            raise ValueError(f"location '{location}' does not exist")
+
+        if not os.path.isdir(location):
+            raise ValueError(f"location '{location}' is not a valid dir")
+    else:
+        location = tempfile.gettempdir()
```

Comment:
if tempdir here, who is responsible for removing it?

Reply:
The caller will get the path of the file after the download is done. The caller can delete the file after that.
In case of tempdir, it is managed by the OS (which periodically cleans it). If the caller remembers to remove the file, then it's okay.